File: cl-arch-mvp/android/core/usecases/CreateUserUseCase.kt
// Dominio: Use Case
data class CreateUserRequest(
    val name: String,
    val email: String
)

data class CreateUserResponse(
    val userId: String
)

interface CreateUserUseCase {
    fun execute(request: CreateUserRequest): CreateUserResponse
}

class CreateUserUseCaseImpl(private val interactor: CreateUserInteractor) : CreateUserUseCase {
    override fun execute(request: CreateUserRequest): CreateUserResponse {
        return interactor.execute(request)
    }
}

// Interactor
class CreateUserInteractor {
    fun execute(request: CreateUserRequest): CreateUserResponse {
        // lógica dummy
        return CreateUserResponse(userId = "123")
    }
}

// Coordinator
object CreateUserCoordinator {
    fun build(): CreateUserUseCase {
        val interactor = CreateUserInteractor()
        return CreateUserUseCaseImpl(interactor)
    }
}

File: cl-arch-mvp/ios/Core/UseCases/CreateUserUseCase.swift
// Dominio: Use Case
struct CreateUserRequest {
    let name: String
    let email: String
}

struct CreateUserResponse {
    let userId: String
}

protocol CreateUserUseCase {
    func execute(request: CreateUserRequest) throws -> CreateUserResponse
}

final class CreateUserUseCaseImpl: CreateUserUseCase {
    private let interactor: CreateUserInteractor
    init(interactor: CreateUserInteractor) {
        self.interactor = interactor
    }
    func execute(request: CreateUserRequest) throws -> CreateUserResponse {
        return try interactor.execute(request: request)
    }
}

// Interactor
class CreateUserInteractor {
    func execute(request: CreateUserRequest) throws -> CreateUserResponse {
        // lógica dummy
        return CreateUserResponse(userId: "123")
    }
}

// Coordinator
class CreateUserCoordinator {
    static func build() -> CreateUserUseCase {
        let interactor = CreateUserInteractor()
        return CreateUserUseCaseImpl(interactor: interactor)
    }
}
